### PR TITLE
FMWK-600 Add support for sorting backup files before reading

### PR DESCRIPTION
**Pull request by filkeith**

### CODE REVIEW

File: io/azure/blob/options.go [hunk 1]
```diff
@@ -123,3 +133,12 @@ func WithSkipDirCheck() Opt {
 		r.skipDirCheck = true
 	}
 }
+
+// WithSorted adds a sorting flag.
+// Which means that files will be read from directory in the sorted order.
+// Is used only for Reader.
+func WithSorted(sort SortOrder) Opt {
```
Comment (reugn) [inline io/azure/blob/options.go:140]: Add validation for the `SortOrder` argument. Panicking is acceptable for unknown values (document this).

File: io/aws/s3/reader.go [hunk 2]
```diff
@@ -141,36 +171,96 @@ func (r *Reader) streamDirectory(
...
 
-				// We check *p.Key == nil in the beginning.
-				errorsCh <- fmt.Errorf("failed to open directory file %s: %w", *p.Key, err)
+	wg.Wait()
+}
 
-				return
-			}
+// sortObjects receives keys, sort them, and then send to open chan.
+func (r *Reader) sortObjects(ctx context.Context, objectsToSort <-chan *string, objectsToOpen chan<- *string) {
```
Comment (reugn) [inline io/aws/s3/reader.go:202]: Following our discussion, consider moving the refactored sorting function to commons.

File: io/aws/s3/options.go [hunk 3]
```diff
@@ -14,6 +14,13 @@
 
 package s3
 
+type SortOrder string
+
+const (
+	SortASC  SortOrder = "asc"
+	SortDESC SortOrder = "desc"
```
[the remark this answers is not in the document]
Comment (filkeith) [reply io/aws/s3/options.go:21]: But how we will use it outside of this package? When we initialize reader `storage.WithSorted(storage.SortASC)`

File: io/aws/s3/reader.go [hunk 4]
```diff
@@ -21,15 +21,20 @@ import (
...
 	"github.com/aws/aws-sdk-go-v2/aws"
 	awsHttp "github.com/aws/aws-sdk-go-v2/aws/transport/http"
 	"github.com/aws/aws-sdk-go-v2/service/s3"
 	"github.com/aws/smithy-go"
 )
 
-const s3type = "s3"
+const (
+	s3type     = "s3"
+	bufferSize = 256
```
[the remark this answers is not in the document]
Comment (filkeith) [reply io/aws/s3/reader.go:36]: add WithBufferSize() method?

File: io/aws/s3/reader.go (hunk 2 above)
Comment (filkeith) [reply io/aws/s3/reader.go:202]: I didn't want to tied up to external package

File: io/aws/s3/options.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (reugn) [reply io/aws/s3/options.go:21]: It will be still exported.

File: io/aws/s3/reader.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (reugn) [reply io/aws/s3/reader.go:36]: Let's leave it for now as is.

File: io/aws/s3/options.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (filkeith) [reply io/aws/s3/options.go:21]: oh, you mean `values` ok

[the remark this answers is not in the document]
Comment (filkeith) [reply io/aws/s3/options.go:21]: oh, my bad, ok

File: io/azure/blob/options.go (hunk 1 above)
Comment (filkeith) [reply io/azure/blob/options.go:140]: added in reader initialization

**Review by reugn — APPROVED** [review]

**Pull request merged**